### PR TITLE
[mlir][spirv] Integrate `convert-to-spirv` into `mlir-vulkan-runner`

### PR DESCRIPTION
**Pull request by obtuseangleAZ**
**Description**
This PR adds a new option for `convert-to-spirv` pass to clone and convert only GPU kernel modules for integration testing. The reason for using pass options instead of two separate passes is that they both consist of `memref` types conversion and individual dialect patterns, except they run on different scopes. The PR also replaces the `gpu-to-spirv` pass with the `convert-to-spirv` pass (with the new option)  in `mlir-vulkan-runner`.

**Future Plan**
Use nesting pass pipelines in `mlir-vulkan-runner` instead of adding this option.

### CODE REVIEW

File: mlir/include/mlir/Conversion/Passes.td [hunk 1]
```diff
@@ -50,7 +50,10 @@ def ConvertToSPIRVPass : Pass<"convert-to-spirv"> {
     "Run function signature conversion to convert vector types">,
     Option<"runVectorUnrolling", "run-vector-unrolling", "bool",
     /*default=*/"true",
-    "Run vector unrolling to convert vector types in function bodies">
+    "Run vector unrolling to convert vector types in function bodies">,
+    Option<"runOnGPUModules", "run-on-gpu-modules", "bool",
+    /*default=*/"false",
+    "Clone and convert only the GPU modules for integration testing">
```
Comment (kuhar) [inline mlir/include/mlir/Conversion/Passes.td:56]:
```suggestion
    "Clone and convert GPU modules">
```

This is useable in other contexts,  the pass shouldn't track who the exact users are

File: mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp [hunk 2]
```diff
@@ -64,31 +94,32 @@ struct ConvertToSPIRVPass final
...
-
-    if (failed(applyPartialConversion(op, *target, std::move(patterns))))
-      return signalPassFailure();
+    if (runOnGPUModules) {
+      SmallVector<Operation *, 1> gpuModules;
+      OpBuilder builder(context);
+      op->walk([&](gpu::GPUModuleOp gpuModule) {
+        builder.setInsertionPoint(gpuModule);
+        gpuModules.push_back(builder.clone(*gpuModule));
+      });
```
Comment (kuhar) [inline mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp:103]: Could you explain why we do the cloning here?

File: mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp [hunk 3]
```diff
@@ -64,31 +94,32 @@ struct ConvertToSPIRVPass final
...
+        spirv::TargetEnvAttr targetAttr =
+            spirv::lookupTargetEnvOrDefault(gpuModule);
+        mapToMemRef(gpuModule, targetAttr);
+        populateConvertToSPIRVPatterns(typeConverter, scfToSPIRVContext,
+                                       patterns);
+        if (failed(
+                applyFullConversion(gpuModule, *target, std::move(patterns))))
+          return signalPassFailure();
+      }
+    } else {
```
Comment (kuhar) [inline mlir/lib/Conversion/ConvertToSPIRV/ConvertToSPIRVPass.cpp:116]: Prefer early exits: https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code

File: mlir/include/mlir/Conversion/Passes.td [hunk 4]
```diff
@@ -50,7 +50,10 @@ def ConvertToSPIRVPass : Pass<"convert-to-spirv"> {
     "Run function signature conversion to convert vector types">,
     Option<"runVectorUnrolling", "run-vector-unrolling", "bool",
     /*default=*/"true",
-    "Run vector unrolling to convert vector types in function bodies">
+    "Run vector unrolling to convert vector types in function bodies">,
+    Option<"runOnGPUModules", "run-on-gpu-modules", "bool",
```
Comment (kuhar) [inline mlir/include/mlir/Conversion/Passes.td:54]: Alternative name: `convert-gpu-module`?

**Review by kuhar — COMMENTED** [review]
Could we add a couple of LIt tests for this? Right now the testing depends on the mlir runner, but most folks probably have it disabled and only run cpu tests in their daily development flow.

**Review by kuhar — APPROVED** [review]
Thanks. Having to use a new flag is suboptimal, but overall this seems like a big improvement over using the gpu-to-spirv pass that accumulated unrelated patterns.

**Review by Hardcode84 — APPROVED** [review]

**Comment by joker-eph** [comment]
> Use nesting pass pipelines in mlir-vulkan-runner instead of adding this option.

I'm slightly concerned about seeing any "future plans"  for mlir-vulkan-runner, the status is that this runner should be removed entirely (just like we removed to Cuda runner) to align the testing on mlir-cpu-runner.

It seems to me that it would be more fruitful to concentrate the efforts to actually remove it instead, it's a long-time pending TODO.

**Pull request merged**